src/estimagic/parameters/nonlinear_constraints.py

```diff
@@ -0,0 +1,468 @@
+
+    Args:
+        nonlinear_constraints (list[dict]): List of dictionaries, each representing a
+            nonlinear constraint.
+        params (pandas): A pytree containing the parameters with respect to which the
+            criterion is optimized. Examples are a numpy array, a pandas Series,
+            a DataFrame with "value" column, a float and any kind of (nested) dictionary
+            or list containing these elements. See :ref:`params` for examples.
+        converter (Converter): NamedTuple with methods to convert between internal and
+            external parameters, derivatives and function outputs.
+        numdiff_options (dict): Keyword arguments for the calculation of numerical
+            derivatives. See :ref:`first_derivative` for details. Note that the default
+            method is changed to "forward" for speed reasons. Contains lower and upper
+            bounds of parameters.
+        skip_checks (bool): Whether checks on the inputs are skipped. This makes the
+            optimization faster, especially for very fast constraint functions. Default
+            False.
+
+    Returns:
+        list[dict]: List of processed constraints.
+
+    """
+    constraint_evals = [
+        _check_validity_and_return_evaluation(c, params, skip_checks)
+        for c in nonlinear_constraints
+    ]
+
+    _processor = partial(
+        _process_nonlinear_constraint,
+        params=params,
+        converter=converter,
+        numdiff_options=numdiff_options,
+    )
+
+    processed = [
+        _processor(c, e) for c, e in zip(nonlinear_constraints, constraint_evals)
+    ]
+    return processed
+
+
+def _process_nonlinear_constraint(
+    c, constraint_eval, params, converter, numdiff_options
+):
+    """Process a single nonlinear constraint."""
+
+    # ==================================================================================
+    # Process selectors for params (functional) and flat_params (indices)
+    # ==================================================================================
+
+    external_selector = _process_selector(c)  # functional selector
+
+    selection_indices, n_params = _get_selection_indices(params, external_selector)
+
+    # ==================================================================================
+    # Evaluate constraint function and selector
+    # ==================================================================================
+    selected = external_selector(params)
+
+    constraint_func = c["func"]
+
+    if constraint_eval is None:
+        constraint_eval = constraint_func(selected)
+
+    # ==================================================================================
+    # Retrieve number of constraints
+    # ==================================================================================
+    if "n_constr" in c:
+        _n_constr = c["n_constr"]
+    else:
+        constraint_value = constraint_func(external_selector(params))
```

Comment:
Isn't this evaluation redundant if `constraint_eval` is not `None`? Would a user ever want to specify `n_constr` if it only buys one less evaluation of the constraint? For least squares optimizers we chose not to support things like `n_residuals` because we thought that the gains are small and the error potential large.